### PR TITLE
Add mangling support for `@differentiable` function types.

### PR DESCRIPTION
**Pull request by dan-zheng**
Add mangling support for `@differentiable` and `@differentiable(linear)`
function types.

This fixes debug info builds. Previously, IRGenDebugInfo crashed because
`@differentiable` function types did not have full mangling support.
`IRGenDebugInfoImpl::getOrCreateType` computes `llvm::DIType` type debug
info by demangling type names, leading to a type size inconsistency for
`@differentiable` function types.

Change `tensorflow_osx_base` build preset to use `--release-debuginfo`.
macOS toolchains (with TensorFlow) now have debug info again.

Resolves [TF-597](https://bugs.swift.org/projects/TF/issues/TF-597).

### CODE REVIEW

**Comment by dan-zheng** [comment]
Thank you @compnerd for spending hours to debug this together - it was a blast 🙂

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by rxwei** [comment]
Thank you @compnerd and @dan-zheng !

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by eaplatanios** [comment]
That's great! Thanks @dan-zheng and @compnerd !

Is it possible that this allows us to now build the `TensorFlow` module with optimizations enabled? If I recall correctly, the error when enabling optimizations previously was related to IRGen and auto-diff.

File: test/SILGen/mangling.swift [hunk 1]
```diff
@@ -184,3 +184,11 @@ func varargsVsArray(arr: [Int], n: String) { }
 
 // CHECK-LABEL: sil hidden [ossa] @$s8mangling14varargsVsArray3arr1nySaySiGd_SStF : $@convention(thin) (@guaranteed Array<Array<Int>>, @guaranteed String) -> ()
 func varargsVsArray(arr: [Int]..., n: String) { }
+
+// SWIFT_ENABLE_TENSORFLOW
+// CHECK-LABEL: sil hidden [ossa] @$s8mangling15funcVsDiffFunc12fnyS2fXE_tF : $@convention(thin) (@noescape @callee_guaranteed (Float) -> Float) -> ()
```
Comment (dan-zheng) [inline test/SILGen/mangling.swift]: Note: these newly added tests actually test `@differentiable` function mangling logic added in https://github.com/apple/swift/pull/25804.

That logic is different from the mangling logic added in this patch (mangling modifiers for `ImplFunctionType`). I'm not sure what's the meaning of `ImplFunctionType` and why it exists separately from `SILFunctionType` mangling: why is there `ImplEscaping` (for `ImplFunctionType`) in addition to `EscapingObjCBlock`/`EscapingAutoClosureType` (for regular `SILFunctionType`)?

cc @slavapestov, wonder if you can clarify

**Comment by dan-zheng** [comment]
> Is it possible that this allows us to now build the `TensorFlow` module with optimizations enabled? If I recall correctly, the error when enabling optimizations previously was related to IRGen and auto-diff.

I'm not sure - could you please share the errors from building `TensorFlow` with `-O`?
The patch fixes "compiler crash when compiling `@differentiable`-function-typed parameters with `-g` and `-O`".

File: lib/Demangling/Demangler.cpp [hunk 2]
```diff
@@ -1729,6 +1729,12 @@ NodePointer Demangler::demangleImplFunctionType() {
   if (nextIf('e'))
     type->addChild(createNode(Node::Kind::ImplEscaping), *this);
 
+  // SWIFT_ENABLE_TENSORFLOW
+  if (nextIf('d'))
+    type->addChild(createNode(Node::Kind::ImplDifferentiable), *this);
+  if (nextIf('l'))
+    type->addChild(createNode(Node::Kind::ImplLinear), *this);
+
```
Comment (saeta) [inline lib/Demangling/Demangler.cpp]: Swift enable tensorflow end?

**Review by saeta — COMMENTED** [review]
Nice work @dan-zheng !!

File: lib/Demangling/Demangler.cpp (hunk 2 above)
Comment (dan-zheng) [reply lib/Demangling/Demangler.cpp]: There's no strict rule for `// SWIFT_ENABLE_TENSORFLOW END`, I add it only when I deem the "end of added code" to be ambiguous.

File: docs/ABI/Mangling.rst [hunk 3]
```diff
@@ -583,12 +583,16 @@ mangled in to disambiguate.
   impl-function-type ::= type* 'I' FUNC-ATTRIBUTES '_'
   impl-function-type ::= type* generic-signature 'I' PSEUDO-GENERIC? FUNC-ATTRIBUTES '_'
 
-  FUNC-ATTRIBUTES ::= CALLEE-ESCAPE? CALLEE-CONVENTION FUNC-REPRESENTATION? PARAM-CONVENTION* RESULT-CONVENTION* ('z' RESULT-CONVENTION)
+  FUNC-ATTRIBUTES ::= CALLEE-ESCAPE? DIFFERENTIABILITY-KIND? CALLEE-CONVENTION FUNC-REPRESENTATION? PARAM-CONVENTION* RESULT-CONVENTION* ('z' RESULT-CONVENTION)
```
Comment (dan-zheng) [inline docs/ABI/Mangling.rst]: Minor: missing `SWIFT_ENABLE_TENSORFLOW` markers for changes to `docs/ABI/Mangling.rst` added in this patch and in https://github.com/apple/swift/pull/25804. I'm not sure how to add comments in `.rst` files though.

**Comment by eaplatanios** [comment]
@dan-zheng I'll test after this PR is merged because lots of changes have happened since last time I attempted enabling optimizations.

File: docs/ABI/Mangling.rst (hunk 3 above)
Comment (rxwei) [reply docs/ABI/Mangling.rst]: Comments start with `..` I believe. We have `.. SWIFT_ENABLE_TENSORFLOW` in SIL.rst.

Comment (dan-zheng) [reply docs/ABI/Mangling.rst]: Thanks! Done in https://github.com/apple/swift/pull/26597.

File: lib/Demangling/OldRemangler.cpp [hunk 4]
```diff
@@ -1267,6 +1267,18 @@ void Remangler::mangleImplResult(Node *node) {
...
 
+// SWIFT_ENABLE_TENSORFLOW
+void Remangler::mangleImplDifferentiable(Node *node) {
+  // The old mangler does not encode `@differentiable` function types.
+  Buffer << 'd';
+}
+
+// SWIFT_ENABLE_TENSORFLOW
+void Remangler::mangleImplLinear(Node *node) {
+  // The old mangler does not encode `@differentiable(linear)` function types.
```
Comment (compnerd) [inline lib/Demangling/OldRemangler.cpp]: The comment doesn't really match the code.

File: lib/Demangling/OldRemangler.cpp [hunk 5]
```diff
@@ -1267,6 +1267,18 @@ void Remangler::mangleImplResult(Node *node) {
   mangleChildNodes(node); // impl convention, type
 }
 
+// SWIFT_ENABLE_TENSORFLOW
+void Remangler::mangleImplDifferentiable(Node *node) {
+  // The old mangler does not encode `@differentiable` function types.
+  Buffer << 'd';
+}
+
+// SWIFT_ENABLE_TENSORFLOW
```
Comment (compnerd) [inline lib/Demangling/OldRemangler.cpp]: This should be lowered ~5 lines, and is missing an ` END`.

File: lib/Demangling/Remangler.cpp [hunk 6]
```diff
@@ -1379,6 +1379,16 @@ void Remangler::mangleIVarDestroyer(Node *node) {
   Buffer << "fE";
 }
 
+// SWIFT_ENABLE_TENSORFLOW
+void Remangler::mangleImplDifferentiable(Node *node) {
+  Buffer << 'd';
+}
+
+// SWIFT_ENABLE_TENSORFLOW
```
Comment (compnerd) [inline lib/Demangling/Remangler.cpp]: Shift ~4 lines below, missing ` END`.

File: lib/IRGen/GenDiffFunc.cpp [hunk 7]
```diff
@@ -32,6 +32,7 @@
 using namespace swift;
 using namespace irgen;
 
+/// A pair of `@differentiable` function extractee and differentiation order.
```
Comment (compnerd) [inline lib/IRGen/GenDiffFunc.cpp]: Is this related to this change?

File: test/AutoDiff/differentiable_func_debuginfo.swift [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+// to a type size inconsistency.
+//
+// Conclusion: mangling coverage is important.
+
+// Minimal dummy compiler-known `Differentiable` protocol.
+public protocol Differentiable {
+  associatedtype TangentVector
+}
+
+// This declaration is necessary to reproduce the crash for some reason.
```
Comment (compnerd) [inline test/AutoDiff/differentiable_func_debuginfo.swift]: Could you please update this comment?

What is happening here is pretty simple.  `Builtin.autodiffApply_jvp` is constructing a use of the `tf597ProblematicVarDecl` type which would be decorated without the `@differential`.  The parameter for `blackHole` is of type `$@noescape @callee_guaranteed (@in_guaranteed T) -> @out U` which will match the decorated name for the differential function when the `@differential` attribute is not accounted for.  As a result, when generating the debug information the types are uniqued the two values are treated as the same except the type in `blackHole` is smaller than the `@differential` type which causes the validation to fail.

File: test/Demangle/Inputs/manglings.txt [hunk 9]
```diff
@@ -351,3 +351,6 @@ $s18keypaths_inlinable13KeypathStructV8computedSSvpACTKq  ---> key path getter f
 $s18resilient_protocol24ResilientDerivedProtocolPxAA0c4BaseE0Tn --> associated conformance descriptor for resilient_protocol.ResilientDerivedProtocol.A: resilient_protocol.ResilientBaseProtocol
 $s3red4testyAA3ResOyxSayq_GAEs5ErrorAAq_sAFHD1__HCg_GADyxq_GsAFR_r0_lF --> red.test<A, B where B: Swift.Error>(red.Res<A, B>) -> red.Res<A, [B]>
 $s3red4testyAA7OurTypeOy4them05TheirD0Vy5AssocQzGAjE0F8ProtocolAAxAA0c7DerivedH0HD1_AA0c4BaseH0HI1_AieKHA2__HCg_GxmAaLRzlF ---> red.test<A where A: red.OurDerivedProtocol>(A.Type) -> red.OurType<them.TheirType<A.Assoc>>
+// SWIFT_ENABLE_TENSORFLOW
+$sxq_Idgnr_D ---> @differentiable @callee_guaranteed (@in_guaranteed A) -> (@out B)
+$sxq_Ilgnr_D ---> @differentiable(linear) @callee_guaranteed (@in_guaranteed A) -> (@out B)
```
Comment (compnerd) [inline test/Demangle/Inputs/manglings.txt]: Missing `SWIFT_ENABLE_TENSORFLOW END`?

File: test/SILGen/mangling.swift [hunk 10]
```diff
@@ -184,3 +184,11 @@ func varargsVsArray(arr: [Int], n: String) { }
...
+func funcVsDiffFunc1(fn: (Float) -> Float) {}
+
+// SWIFT_ENABLE_TENSORFLOW
```
Comment (compnerd) [inline test/SILGen/mangling.swift]: Should this be lowered ~3 lines and have an ` END`?

File: lib/Demangling/OldRemangler.cpp (hunk 5 above)
Comment (dan-zheng) [reply lib/Demangling/OldRemangler.cpp]: > This should be lowered ~5 lines,

Are you optimizing the location of these code blocks? I put the code blocks in the order that nodes are defined in `include/swift/Demangling/DemangleNodes.def`.

We haven't been religiously adding `SWIFT_ENABLE_TENSORFLOW END` in all cases, only when the end of the newly added code is ambiguous. I suppose we could start doing so - that requires some team communication.

File: lib/Demangling/OldRemangler.cpp (hunk 4 above)
Comment (dan-zheng) [reply lib/Demangling/OldRemangler.cpp]: Oops, that's a good point. I'm not sure when this code path (OldRemangler) is even triggered.
I'd like to be able to add a test for this code path (whether the test checks that remangling occurs/doesn't occur) before changing this OldRemangler comment/code.

File: lib/Demangling/OldRemangler.cpp (hunk 5 above)
Comment (dan-zheng) [reply lib/Demangling/OldRemangler.cpp]: > This should be lowered ~5 lines

Oh, I understand now. You would like to group the entire code block within `SWIFT_ENABLE_TENSORFLOW` and `SWIFT_ENABLE_TENSORFLOW END`. That would be a better organization of these comments, but we haven't been following strict rules for these comments.

File: lib/Demangling/Remangler.cpp (hunk 6 above)
Comment (dan-zheng) [reply lib/Demangling/Remangler.cpp]: From above:
> That would be a better organization of these comments, but we haven't been following strict rules for these comments.

Optimizing/polishing these comments could be satisfying but they're not very important. We hope to upstream some of this code soon, too.

File: lib/IRGen/GenDiffFunc.cpp (hunk 7 above)
Comment (dan-zheng) [reply lib/IRGen/GenDiffFunc.cpp]: No, but I felt it was okay to add a not-exactly-related doc comment.

I added this comment because it clarifies `fields.push_back(std::make_pair(AutoDiffFunctionExtractInst::Extractee::VJP, 1));` in [`TypeConverter::convertDifferentiableFunctionType`](https://github.com/apple/swift/blob/51dc28d6266d21e2f27c4c600a1f35f32654d069/lib/IRGen/GenDiffFunc.cpp#L185). We came across this while debugging the IRGenDebugInfo crash.

File: test/AutoDiff/differentiable_func_debuginfo.swift (hunk 8 above)
Comment (dan-zheng) [reply test/AutoDiff/differentiable_func_debuginfo.swift]: That makes sense, thanks!

Comment (dan-zheng) [reply test/AutoDiff/differentiable_func_debuginfo.swift]: Done in https://github.com/apple/swift/pull/26600.

File: lib/Demangling/Remangler.cpp (hunk 6 above)
Comment (dan-zheng) [reply lib/Demangling/Remangler.cpp]: Optimized comment in https://github.com/apple/swift/pull/26600.

File: lib/Demangling/OldRemangler.cpp (hunk 5 above)
Comment (dan-zheng) [reply lib/Demangling/OldRemangler.cpp]: Optimized comment in https://github.com/apple/swift/pull/26600.

File: lib/Demangling/OldRemangler.cpp (hunk 4 above)
Comment (dan-zheng) [reply lib/Demangling/OldRemangler.cpp]: Fixed comments in https://github.com/apple/swift/pull/26600. Linked to [TF-750](https://bugs.swift.org/browse/TF-750).

File: test/Demangle/Inputs/manglings.txt (hunk 9 above)
Comment (dan-zheng) [reply test/Demangle/Inputs/manglings.txt]: Added in https://github.com/apple/swift/pull/26600/commits/f2ca6bc5591c0b09b7c8ea3de4e9bad67b007885.

File: test/SILGen/mangling.swift (hunk 10 above)
Comment (dan-zheng) [reply test/SILGen/mangling.swift]: Done in https://github.com/apple/swift/pull/26600/commits/f2ca6bc5591c0b09b7c8ea3de4e9bad67b007885.
